src/plugins/intel_cpu/src/emitters/x64/jit_snippets_emitters.cpp

```diff
@@ -1398,22 +1402,28 @@ void HorizonSumEmitter::emit_isa(const std::vector<size_t> &in, const std::vecto
             Xmm, isa == dnnl::impl::cpu::x64::avx2, Ymm, Zmm>::type;
 
     Vmm src_vmm = Vmm(in[0]);
-    Xmm dst_xmm = Xmm(out[0]);
-    Xmm aux_xmm = Xmm(aux_vec_idxs[0]);
-
-    Reg64 aux_reg = Reg64(aux_gpr_idxs[0]);
-
-    const size_t vlen = dnnl::impl::cpu::x64::cpu_isa_traits<isa>::vlen;
-    const size_t vec_size = vlen / sizeof(float);
-    h->sub(h->rsp, vlen);
-    h->uni_vmovups(h->ptr[h->rsp], src_vmm);
-    h->uni_vpxor(dst_xmm, dst_xmm, dst_xmm);
-    for (size_t i = 0; i < vec_size; i++) {
-        h->mov(aux_reg, h->ptr[h->rsp + i * sizeof(float)]);
-        h->vmovq(aux_xmm, aux_reg);
-        h->uni_vaddps(dst_xmm, dst_xmm, aux_xmm);
+    Vmm dst_vmm = Vmm(out[0]);
+    Vmm aux_vmm = Vmm(aux_vec_idxs[0]);
+
+    if (in[0] != out[0])
+        h->uni_vmovups(dst_vmm, src_vmm);
+    if (isa == dnnl::impl::cpu::x64::avx512_core) {
+        Zmm dst_zmm = Zmm(out[0]);
+        Zmm aux_zmm = Zmm(aux_vec_idxs[0]);
+        h->vshuff32x4(aux_zmm, dst_zmm, dst_zmm, 0x4E);
+        h->uni_vaddps(dst_zmm, dst_zmm, aux_zmm);
```

Comment:
Seems like we can implement base class `HorizonEmitter` with virtual method `perform_op`:

- `HorizonEmitter::HorizonSumEmitter::perform_op(...) { h->uni_vaddps(...); } `
- `HorizonEmitter::HorizonMaxEmitter::perform_op(...) { h->uni_vmaxps(...); } `
- `HorizonEmitter::HorizonYEmitter::perform_op(...) { h->uni_vYps(...); } `

 `HorizonEmitter` has the main common reduction method which calculates reduction (reduction algorithm is common for any ops) and call virtual method `perform_op`.
 
 Overall, `HorizonEmitter` has common reduction logic where each derived class provides the own math operation using `perform_op`. It'd be really easy to add support for other reduction ops in the near future
 
 Could you please think about his idea? Can we implement so? Thank you very much!

Reply:
Great idea! Applied. Thanks Alexandra!